pkg/distributor/validate.go

```diff
@@ -481,9 +497,29 @@ func validateLabels(m *sampleValidationMetrics, cfg labelValidationConfig, userI
 
 		lastLabelName = l.Name
 	}
+
 	return nil
 }
 
+func hashLabelValueInto(dst, src mimirpb.UnsafeMutableString) mimirpb.UnsafeMutableString {
+	h := blake2b.Sum256(unsafeMutableStringToBytes(src))
+
+	buf := unsafeMutableStringToBytes(dst)
+	// Encode as hex inline instead of fmt.Sprintf to avoid allocations due to interface values.
+	copy(buf, "(hash:")
+	const hexChars = "0123456789abcdef"
+	for i, b := range h[:8] {
+		buf[6+i*2] = hexChars[b>>4]
+		buf[7+i*2] = hexChars[b&0x0f]
+	}
+	buf[22] = ')'
```

Comment:
I think we should be using the _whole_ hash here. Given that the hash prefix is only 3.5% of the default label value length limit, it shouldn't be problem, right? Otherwise, if you want a shorter hash, I think you should be converting a shorter blake2b hash explicitly instead.

Also, do you think `hex.Encode` will be inefficient?
```suggestion
	_ = hex.Encode(buf[6:], h[:blake2b.Size256])
	buf[6+blake2b.Size256*2] = ')'
```

Reply:
> Also, do you think `hex.Encode` will be inefficient?

Ah, I had somehow missed it!

Done both at [5de3bc7](https://github.com/grafana/mimir/pull/12627/commits/5de3bc7bc3e2de03cd7bc894cb7a1772cea1fa61)